examples/chat/src/models/message.rs

```diff
@@ -0,0 +1,113 @@
+use chrono::NaiveDateTime;
+use diesel::deserialize::FromSqlRow;
+use diesel::dsl::{And, AsSelect, Desc, Eq, Lt};
+use diesel::helper_types::{InnerJoin, Select};
+use diesel::pg::Pg;
+use diesel::prelude::*;
+use diesel::row::Row;
+use diesel::sql_types::Bool;
+use serde::{Deserialize, Serialize};
+use uuid::Uuid;
+
+use super::thread::Thread;
+use super::user::User;
+use crate::schema::messages::{self, dsl as col};
+use crate::schema::{threads, users};
+
+pub type TableWithJoins = InnerJoin<InnerJoin<messages::table, threads::table>, users::table>;
+pub type DefaultSelection = (
+    AsSelect<Message, Pg>,
+    AsSelect<Thread, Pg>,
+    AsSelect<User, Pg>,
+);
+
+#[derive(Associations, Queryable, Selectable, Serialize)]
+#[diesel(belongs_to(Thread))]
+#[diesel(belongs_to(User, foreign_key = author_id))]
+#[diesel(table_name = messages)]
+#[diesel(check_for_backend(Pg))]
+#[serde(rename_all = "camelCase")]
+pub struct Message {
+    id: Uuid,
+
+    body: String,
+
+    #[serde(skip)]
+    author_id: Uuid,
+
+    #[serde(skip)]
+    thread_id: Uuid,
+
+    created_at: NaiveDateTime,
+    updated_at: NaiveDateTime,
+}
+
+#[derive(AsChangeset, Deserialize)]
+#[diesel(table_name = messages)]
+pub struct MessageChangeSet {
+    pub body: String,
+}
+
+#[derive(Deserialize, Insertable)]
+#[diesel(table_name = messages)]
+#[serde(rename_all = "camelCase")]
+pub struct MessageParams {
+    pub body: String,
+
+    pub author_id: Option<Uuid>,
+    pub thread_id: Option<Uuid>,
+}
+
+#[derive(Serialize)]
+pub struct MessageWithJoins {
+    #[serde(flatten)]
+    message: Message,
+    thread: Thread,
```

Comment:
I'm not going to include thread by default in the message joins. The way that the urls are nested implies that the user already has the thread loaded in to memory. I'll definitely load the first 10 or so reactions as a part of the query as well as the total number of reactions - the amount included in the message payload to show the amount of reactions that are below the fold.